src/dispatch/auth/permissons.py

```diff
@@ -0,0 +1,158 @@
+import logging
+from abc import ABC, abstractmethod
+
+from fastapi import HTTPException, Depends
+from starlette.requests import Request
+from starlette.status import HTTP_403_FORBIDDEN
+
+from dispatch.database import get_db
+from dispatch.enums import UserRoles, Visibility
+from dispatch.auth.service import get_current_user
+from dispatch.auth.models import DispatchUser
+from dispatch.incident.models import Incident
+from dispatch.incident import service as incident_service
+
+
+log = logging.getLogger(__name__)
+
+
+class BasePermission(ABC):
+    """
+    Abstract permission that all other Permissions must be inherited from.
+
+    Defines basic error message, status & error codes.
+
+    Upon initialization, calls abstract method  `has_required_permissions`
+    which will be specific to concrete implementation of Permission class.
+
+    You would write your permissions like this:
+
+    .. code-block:: python
+
+        class TeapotUserAgentPermission(BasePermission):
+
+            def has_required_permissions(self, request: Request) -> bool:
+                return request.headers.get('User-Agent') == "Teapot v1.0"
+
+    """
+
+    error_msg = "Forbidden."
```

Comment:
Should we customize this error message based on what the user tries to do?

Reply:
We can override this when we construct the permission. But this is the default.